scene/2d/navigation_obstacle_2d.cpp

```diff
@@ -79,121 +84,63 @@ void NavigationObstacle2D::_bind_methods() {
-#endif // DEBUG_ENABLED
+			_obstacle_enter_tree();
 		} break;
 
 		case NOTIFICATION_EXIT_TREE: {
-			set_physics_process_internal(false);
-			_update_map(RID());
-#ifdef DEBUG_ENABLED
-			RS::get_singleton()->canvas_item_set_parent(debug_canvas_item, RID());
-#endif // DEBUG_ENABLED
-		} break;
-
-		case NOTIFICATION_PAUSED: {
-			if (!can_process()) {
-				map_before_pause = map_current;
-				_update_map(RID());
-			} else if (can_process() && !(map_before_pause == RID())) {
-				_update_map(map_before_pause);
-				map_before_pause = RID();
-			}
-			NavigationServer2D::get_singleton()->obstacle_set_paused(obstacle, !can_process());
+			_obstacle_exit_tree();
 		} break;
 
+		case NOTIFICATION_PAUSED:
 		case NOTIFICATION_UNPAUSED: {
-			if (!can_process()) {
-				map_before_pause = map_current;
-				_update_map(RID());
-			} else if (can_process() && !(map_before_pause == RID())) {
-				_update_map(map_before_pause);
-				map_before_pause = RID();
-			}
 			NavigationServer2D::get_singleton()->obstacle_set_paused(obstacle, !can_process());
 		} break;
 
 		case NOTIFICATION_VISIBILITY_CHANGED: {
 #ifdef DEBUG_ENABLED
-			RS::get_singleton()->canvas_item_set_visible(debug_canvas_item, is_visible_in_tree());
+			if (debug_canvas_item.is_valid()) {
+				RS::get_singleton()->canvas_item_set_visible(debug_canvas_item, is_visible_in_tree());
+			}
 #endif // DEBUG_ENABLED
 		} break;
 
 		case NOTIFICATION_INTERNAL_PHYSICS_PROCESS: {
-			if (is_inside_tree()) {
-				_update_position(get_global_position());
-
-				if (velocity_submitted) {
-					velocity_submitted = false;
-					// only update if there is a noticeable change, else the rvo agent preferred velocity stays the same
-					if (!previous_velocity.is_equal_approx(velocity)) {
-						NavigationServer2D::get_singleton()->obstacle_set_velocity(obstacle, velocity);
-					}
-					previous_velocity = velocity;
-				}
-			}
+			_obstacle_physics_process();
 		} break;
 
 		case NOTIFICATION_DRAW: {
 #ifdef DEBUG_ENABLED
-			if (is_inside_tree()) {
-				bool is_debug_enabled = false;
-				if (Engine::get_singleton()->is_editor_hint()) {
-					is_debug_enabled = true;
-				} else if (NavigationServer2D::get_singleton()->get_debug_enabled() && NavigationServer2D::get_singleton()->get_debug_avoidance_enabled()) {
-					is_debug_enabled = true;
-				}
-
-				if (is_debug_enabled) {
-					RS::get_singleton()->canvas_item_clear(debug_canvas_item);
-					Transform2D debug_transform = Transform2D(0.0, get_global_position());
-					RS::get_singleton()->canvas_item_set_transform(debug_canvas_item, debug_transform);
-					_update_fake_agent_radius_debug();
-					_update_static_obstacle_debug();
-				}
-			}
+			_obstacle_debug_update();
 #endif // DEBUG_ENABLED
 		} break;
 	}
 }
 
 NavigationObstacle2D::NavigationObstacle2D() {
-	obstacle = NavigationServer2D::get_singleton()->obstacle_create();
+	NavigationServer2D *ns2d = NavigationServer2D::get_singleton();
```

Comment:
perhaps `server` instead of `ns2d`? or `navigation_server`?

Reply:
For this `ns2d` was already used in the engine, e.g. tilemap, that is why it is used here.